### PR TITLE
Cop 12035/vulnerabilities fix form builder

### PR DESCRIPTION
**Pull request by hindumathy-cic**
Node upgrade and docker file revised. 
Vulnerabilities fix for form-builder

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM node:lts-alpine as build
+FROM node:16.17.0-alpine as build
```
Comment (aliktb) [inline Dockerfile:1]: Definitely a good idea moving to a fixed version of node as opposed to `lts` which inevitably breaks things every few months

Comment (aliktb) [reply Dockerfile:1]: just curious if you have had any issues with this node version. I've ran into issues with npm 8.11 so I've reverted to node 16.14 which includes npm v 8.5. If there's no issues on this app then I guess it's fine.

File: client/package.json [hunk 2]
```diff
@@ -18,7 +18,7 @@
   },
   "husky": {
     "hooks": {
-      "pre-push": "CI=true react-scripts test"
+      "pre-push": "CI=true react-app-rewired test"
```
Comment (aliktb) [inline client/package.json:21]: just curious what is wrong with `react-scripts`?

File: client/src/config/appConfig.json [hunk 3]
```diff
@@ -1,176 +1,124 @@
...
-          "username": "me@lodev.xyz",
-          "password": "secret"
-        },
         "keycloak": {
-          "tokenUrl": "http://192.168.86.39:8080/auth/realms/elf/protocol/openid-connect/token",
-          "clientId": "form-builder",
-          "secret": "617143cf-cac0-44aa-9a5a-df5382280218"
+          "tokenUrl": "https://sso-dev.notprod.homeoffice.gov.uk/auth/realms/cop-local/protocol/openid-connect/token",
+          "clientId": "formbuilderservice",
+          "secret": "0ffa960a-759c-4e7f-985b-4cc8c6172b89"
```
Comment (aliktb) [inline client/src/config/appConfig.json:12]: I wonder, do these need to be environment variables or is it ok to have them in `appConfig.json`?

File: server/src/controller/PromotionController.ts [hunk 4]
```diff
@@ -45,7 +46,7 @@ export class PromotionController extends BaseHttpController {
             });
             const formToPromoteId = formToPromote.id;
             logger.info(`Checking if ${formToPromoteId} has any subforms...`);
-            const headers = {
+            const headers: any = {
```
Comment (aliktb) [inline server/src/controller/PromotionController.ts:49]: I assume with Typescript 4, it was complaining about missing datatypes

**Review by aliktb — COMMENTED** [review]
just a few comments

File: Dockerfile (hunk 1 above)
Comment (hindumathy-cic) [reply Dockerfile:1]: Application did not have any issues but the dependent packages needed upgrade

File: client/package.json (hunk 2 above)
Comment (hindumathy-cic) [reply client/package.json:21]: I faced this issue https://github.com/facebook/create-react-app/issues/9870#issuecomment-758560474 and used the solution described in this github issues page

File: client/src/config/appConfig.json (hunk 3 above)
Comment (hindumathy-cic) [reply client/src/config/appConfig.json:12]: I haven't checked that

File: server/src/controller/PromotionController.ts (hunk 4 above)
Comment (hindumathy-cic) [reply server/src/controller/PromotionController.ts:49]: image
I faced the above issue and fixed it by making headers as 'any' datatype

**Review by aliktb — APPROVED** [review]
Approving changes

**Pull request merged**